File: 7.二叉树/32.将有序数组转换为二叉搜索树/108.将有序数组转换为二叉搜索树.cpp
/*
 * @lc app=leetcode.cn id=108 lang=cpp
 *
 * [108] 将有序数组转换为二叉搜索树
 */
#include <bits/stdc++.h>
#include <ranges>
using namespace std;
struct TreeNode
{
    int val;
    TreeNode* left;
    TreeNode* right;
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
};
// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* sortedArrayToBST(vector<int>& nums) {
        TreeNode* node = dfs(nums, 0, nums.size() - 1);
        return node;
    }
private:
    TreeNode* dfs(vector<int>& nums, int l, int r)
    {
        if (l > r) return NULL;
        int mid = l + r >> 1;
        TreeNode* cur = new TreeNode(nums[mid]);
        cur->left = dfs(nums, l, mid - 1);
        cur->right = dfs(nums, mid + 1, r);
        return cur;
    }
};
// @lc code=end

File: 面经150/16.分治/108.将有序数组转换为二叉搜索树.py
#
# @lc app=leetcode.cn id=108 lang=python3
#
# [108] 将有序数组转换为二叉搜索树
#
from mytools import *
# @lc code=start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def sortedArrayToBST(self, nums: List[int]) -> Optional[TreeNode]:
        def dfs(l, r):
            if l > r: return None
            mid = l + r >> 1
            x = nums[mid]
            cur = TreeNode(x)
            ll, lr = l, mid - 1
            rl, rr = mid + 1, r
            cur.left = dfs(ll, lr)
            cur.right = dfs(rl, rr)
            return cur
        return dfs(0, len(nums) - 1)
# @lc code=end

